driver/src/main/java/org/neo4j/driver/internal/handlers/pulln/BasicPullResponseHandler.java

```diff
@@ -18,38 +18,262 @@
  */
 package org.neo4j.driver.internal.handlers.pulln;
 
-import org.reactivestreams.Subscription;
-
+import java.util.Map;
 import java.util.function.BiConsumer;
 
-import org.neo4j.driver.internal.spi.ResponseHandler;
 import org.neo4j.driver.Record;
+import org.neo4j.driver.Statement;
+import org.neo4j.driver.Value;
+import org.neo4j.driver.internal.InternalRecord;
+import org.neo4j.driver.internal.handlers.PullResponseCompletionListener;
+import org.neo4j.driver.internal.handlers.RunResponseHandler;
+import org.neo4j.driver.internal.messaging.request.PullMessage;
+import org.neo4j.driver.internal.spi.Connection;
+import org.neo4j.driver.internal.util.MetadataExtractor;
+import org.neo4j.driver.internal.value.BooleanValue;
 import org.neo4j.driver.summary.ResultSummary;
 
-public interface BasicPullResponseHandler extends ResponseHandler, Subscription
+import static java.lang.String.format;
+import static java.util.Collections.emptyMap;
+import static java.util.Objects.requireNonNull;
+import static org.neo4j.driver.internal.handlers.pulln.FetchSizeUtil.UNLIMITED_FETCH_SIZE;
+import static org.neo4j.driver.internal.messaging.request.DiscardMessage.newDiscardAllMessage;
+
+/**
+ * In this class we have a hidden state machine.
+ * Here is how it looks like:
+ * |                    | DONE | FAILED | STREAMING                      | READY              | CANCELED       |
+ * |--------------------|------|--------|--------------------------------|--------------------|----------------|
+ * | request            | X    | X      | toRequest++ ->STREAMING        | PULL ->STREAMING   | X              |
+ * | cancel             | X    | X      | ->CANCELED                     | DISCARD ->CANCELED | ->CANCELED     |
+ * | onSuccess has_more | X    | X      | ->READY request if toRequest>0 | X                  | ->READY cancel |
+ * | onSuccess          | X    | X      | summary ->DONE                 | X                  | summary ->DONE |
+ * | onRecord           | X    | X      | yield record ->STREAMING       | X                  | ->CANCELED     |
+ * | onFailure          | X    | X      | ->FAILED                       | X                  | ->FAILED       |
+ *
+ * Currently the error state (marked with X on the table above) might not be enforced.
```

Comment:
`DONE` is an error state?

Reply:
If it is done, then there is no further transition you can go from Done state. 

The top row of the table defines what state you are currently in. Then the first column defines what action you can perform.

I know this is a bit shaky. I will add a card to improve this class with a proper state machine as you pointed out.